File: src/main/java/eu/happybe/openapi/friends/FriendsManager.java
package eu.happybe.openapi.friends;

import eu.happybe.openapi.api.form.FormQueue;
import eu.happybe.openapi.api.form.types.ModalForm;
import eu.happybe.openapi.mysql.QueryQueue;
import eu.happybe.openapi.mysql.query.AddFriendQuery;
import io.gomint.GoMint;
import io.gomint.entity.EntityPlayer;

import java.util.function.BiConsumer;

public class FriendsManager {

    public static void setFriends(EntityPlayer player, EntityPlayer friend) {
        FriendsManager.setFriends(player, friend, null);
    }

    public static void setFriends(EntityPlayer player, EntityPlayer friend, BiConsumer<EntityPlayer, AddFriendQuery> callback) {
        QueryQueue.submitQuery(new AddFriendQuery(player.name(), friend.name()), addFriendsQuery -> {
            if(callback != null) {
                callback.accept(player, (AddFriendQuery) addFriendsQuery);
            }
        });
    }

    // TODO - Move this to BasicEssentials
    public static void sendFriendRequest(EntityPlayer player, EntityPlayer newFriend) {
        ModalForm form = new ModalForm("Friend Request", player.name() + " sent you a friend request.");
        form.setFirstButton("§aAccept");
        form.setSecondButton("§cDecline");

        form.setCustomData(player.name());

        form.setCallable(response -> {
            EntityPlayer friend = response.getPlayer();
            if(response.getButtonClicked() == 1) {
                friend.sendMessage("§9Friends> §cFriend request cancelled.");
                return;
            }

            EntityPlayer sender = GoMint.instance().findPlayerByName((String)response.getForm().getCustomData());
//            EntityPlayer sender = Server.getInstance().getEntityPlayerExact((String) response.getForm().getCustomData());
            if(sender == null || !sender.online()) {
                friend.sendMessage("§9Friends> §cFriend request expired.");
                return;
            }

            sender.sendMessage("§9Friends> §a" + friend + " accepted your friend request.");
            friend.sendMessage("§9Friends> §aFriend request accepted!");

            FriendsManager.setFriends(sender, friend);
        });

        FormQueue.sendForm(newFriend, form);
    }
}

File: src/main/java/eu/happybe/openapi/friends/FriendsManager.kt
package eu.happybe.openapi.friends

import cn.nukkit.Player
import cn.nukkit.Server
import eu.happybe.openapi.form.FormCallbackResponse
import eu.happybe.openapi.form.FormQueue
import eu.happybe.openapi.form.types.ModalForm
import eu.happybe.openapi.mysql.AsyncQuery
import eu.happybe.openapi.mysql.QueryQueue
import eu.happybe.openapi.mysql.query.AddFriendQuery
import java.util.function.BiConsumer

object FriendsManager {
    fun setFriends(player: Player, friend: Player) {
        setFriends(player, friend, null)
    }

    fun setFriends(player: Player, friend: Player, callback: BiConsumer<Player?, AddFriendQuery?>?) {
        QueryQueue.submitQuery(AddFriendQuery(player.name, friend.name)) { addFriendsQuery: AsyncQuery? -> callback?.accept(player, addFriendsQuery as AddFriendQuery?) }
    }

    // TODO - Move this to BasicEssentials
    fun sendFriendRequest(player: Player, newFriend: Player) {
        val form = ModalForm("Friend Request", player.name + " sent you a friend request.")
        form.setFirstButton("§aAccept")
        form.setSecondButton("§cDecline")
        form.customData = player.name
        form.setCallable { response: FormCallbackResponse? ->
            val friend = response.getPlayer()
            if (response.getButtonClicked() == 1) {
                friend.sendMessage("§9Friends> §cFriend request cancelled.")
                return@setCallable
            }
            val sender = Server.getInstance().getPlayerExact(response.getForm().customData as String)
            if (sender == null || !sender.isOnline) {
                friend.sendMessage("§9Friends> §cFriend request expired.")
                return@setCallable
            }
            sender.sendMessage("§9Friends> §a$friend accepted your friend request.")
            friend.sendMessage("§9Friends> §aFriend request accepted!")
            setFriends(sender, friend)
        }
        FormQueue.sendForm(newFriend, form)
    }
}